nemo/deploy/nlp/megatronllm_deployable.py

```diff
@@ -0,0 +1,321 @@
+    """
+    return (-1 if type(pyvalue) == list else 1,)
+
+
+def GetNumpyDtype(pyvalue):
+    """
+    utility function to get numpy dtype of a python value
+    e.g. bool -> np.bool_
+    """
+    '''
+    manually defining the mapping of python type -> numpy type for now
+    is there a better way to do it?  tried np.array(pyvalue).dtype, but that doesn't seem to work
+    '''
+    py_to_numpy_mapping = {str: bytes, bool: np.bool_, float: np.single, int: np.int_}
+    python_type = type(pyvalue)
+    # for lists, return the type of the internal elements
+    if python_type == list:
+        python_type = type(pyvalue[0])
+    numpy_type = py_to_numpy_mapping[python_type]
+    return numpy_type
+
+
+class ServerSync(IntEnum):
+    """Enum for synchronization messages using torch.distributed"""
+
+    WAIT = auto()
+    SIGNAL = auto()
+
+    def to_long_tensor(self):
+        return torch.tensor([self], dtype=torch.long, device='cuda')
+
+
+class MegatronLLMDeployable(ITritonDeployable):
+    """Triton inference server compatible deploy class for a .nemo model file"""
+
+    def __init__(
+        self,
+        nemo_checkpoint_filepath: str = None,
+        num_devices: int = 1,
+        num_nodes: int = 1,
+        existing_model: MegatronGPTModel = None,
+    ):
+        if nemo_checkpoint_filepath is None and existing_model is None:
+            raise ValueError(
+                "MegatronLLMDeployable requires either a .nemo checkpoint filepath or an existing MegatronGPTModel, but both provided were None"
+            )
+        if num_devices > 1:
+            LOGGER.warning(
+                "Creating a MegatronLLMDeployable with num_devices>1 will assume running with a PyTorch Lightning DDP-variant strategy, which will run the main script once per device. Make sure any user code is compatible with multiple executions!"
+            )
+
+        # if both existing_model and nemo_checkpoint_filepath are provided, existing_model will take precedence
+        if existing_model is not None:
+            self.model = existing_model
+        else:
+            self._load_from_nemo_checkpoint(nemo_checkpoint_filepath, num_devices, num_nodes)
+
+        self.model.eval()
+        # helper threads spawned by torch.multiprocessing should loop inside this helper function
+        self._helper_thread_evaluation_loop()
+
+    def _load_from_nemo_checkpoint(self, nemo_checkpoint_filepath: str, num_devices: int, num_nodes: int):
+        if Path(nemo_checkpoint_filepath).exists():
+            trainer = Trainer(
+                strategy=NLPDDPStrategy(),
+                devices=num_devices,
+                num_nodes=num_nodes,
+            )
+
+            custom_config = MegatronGPTModel.restore_from(
+                nemo_checkpoint_filepath, trainer=trainer, return_config=True
+            )
+            # transformer_engine should always be true according to EricH, but GPT-2B model will fail if it is enabled
```

Comment:
Let's remove the `according to EricH, but GPT-2B model will fail if it is enabled` part and assume that it should be set to true for now.